Add an action to review add-ons in the store

nvdaes opened the pull request:
- Addan action to review add-ons in the store
- Fix linter error and add changelog

<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review". 
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
Closes #15576
### Summary of the issue:
The add-on store doesn't have a mechanism to see and provide feedback for add-ons.
### Description of user facing changes
- Now, it's possible to open a webpage to provide or see feedback for add-ons from an action added to the add-on store.
- The review URL is shown in the details panel for add-ons including a reviewUrl in metadata.
### Description of development approach
An action has been added to provide feedback for add-ons, in a similar way to the approach followed for the license URL. Also, this has been added to the details panel.
### Testing strategy:
- Tested manually in English to ensure that "f" shortcut doesn't produce conflicts with other shortcuts in the actions menu.
- Ensure that the action is available for Alzaker add-on, which includes a reviewUrl in metadata. Also, ensure that the review URL is shown in the details panel.
### Known issues with pull request:
None
### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

nvdaes committed:
Addan action to review add-ons in the store

nvdaes committed:
Fix linter error and add changelog

Adriani90 commented:
Thanks Noelia for this great work! I think this action and the feedback process from submiting to how the feedback reaches the addon author definitely diserves a sub chapter in the user guide.

nvdaes commented:
Thank you @Adriani90 . Let's see how to make a subchapter in the user guide. It's not easy since actions aren't described in detail. I'll try to do it.

nvdaes committed:
Merge branch 'reviewUrlAction' of https://github.com/nvdaes/nvda into reviewUrlAction

nvdaes committed:
Add subchapter in the user guide

XLTechie commented on user_docs/en/userGuide.t2t:
```diff
@@ -2840,6 +2840,13 @@ If the add-on was previously "disabled", the status will show "enabled after res
 If the add-on was previously "enabled", the status will show "disabled after restart".
 Just like when you install or remove add-ons, you need to restart NVDA in order for changes to take effect.
 
++++ Reviewing and consulting feedback for add-ons +++[AddonStoreReviews]
```
This is okay, but an alternative to consider:

```suggestion
+++ Reviewing add-ons and reading reviews +++[AddonStoreReviews]
```

XLTechie commented on user_docs/en/userGuide.t2t:
```diff
@@ -2840,6 +2840,13 @@ If the add-on was previously "disabled", the status will show "enabled after res
 Just like when you install or remove add-ons, you need to restart NVDA in order for changes to take effect.
 
++++ Reviewing and consulting feedback for add-ons +++[AddonStoreReviews]
+Before installing an add-on, you may want to see opinions and reactions by others.
```
```suggestion
Before installing an add-on, you may want to read opinions and reactions by others.
```

XLTechie commented on user_docs/en/userGuide.t2t:
```diff
@@ -2840,6 +2840,13 @@ If the add-on was previously "disabled", the status will show "enabled after res
 
++++ Reviewing and consulting feedback for add-ons +++[AddonStoreReviews]
+Before installing an add-on, you may want to see opinions and reactions by others.
+Also, you may want to provide feedback about add-ons to help other users.
```
```suggestion
Also, it is helpful to other users, if you  take time to provide feedback about add-ons you have tried.
```

XLTechie commented on user_docs/en/userGuide.t2t:
```diff
@@ -2840,6 +2840,13 @@ If the add-on was previously "disabled", the status will show "enabled after res
++++ Reviewing and consulting feedback for add-ons +++[AddonStoreReviews]
+Before installing an add-on, you may want to see opinions and reactions by others.
+Also, you may want to provide feedback about add-ons to help other users.
+To see or share your feedback about add-ons, select an add-on and use the See or provide feedback option in the actions menu.
```
I really do think that "read" is a better word than "see" for both the option and the usage. Not for any "ableist language" reason, but just because it flows better in English IMO.

Also, for this one, you are not going to "see your feedback", or even "read your feedback"--you are going to either read others' feedback, or leave your own.

Lastly, I think the menu option should be shorter--perhaps just "Reviews". I have only changed the word "see" in the suggestion though.

```suggestion
To read what others have said, or share your feedback about add-ons, select an add-on and use the Read or provide feedback option in the actions menu.
```

XLTechie commented on user_docs/en/userGuide.t2t:
```diff
@@ -2840,6 +2840,13 @@ If the add-on was previously "disabled", the status will show "enabled after res
+Before installing an add-on, you may want to see opinions and reactions by others.
+Also, you may want to provide feedback about add-ons to help other users.
+To see or share your feedback about add-ons, select an add-on and use the See or provide feedback option in the actions menu.
+This will open a dedicated webpage at GitHub Discussions, where you will be able to see or share your reactions for the add-on, or even for specific versions.
```
Again, you probably aren't going to go there to see your own reactions, especially if you haven't written them yet. I played with the language a bit in this one.

```suggestion
This will open a dedicated webpage at GitHub Discussions, where you will be able to read and write reviews for the add-on, or even comment on specific versions.
```

XLTechie commented on user_docs/en/userGuide.t2t:
```diff
@@ -2840,6 +2840,13 @@ If the add-on was previously "disabled", the status will show "enabled after res
+Also, you may want to provide feedback about add-ons to help other users.
+To see or share your feedback about add-ons, select an add-on and use the See or provide feedback option in the actions menu.
+This will open a dedicated webpage at GitHub Discussions, where you will be able to see or share your reactions for the add-on, or even for specific versions.
+Please, keep account that this doesn't replace direct communication with add-on developers. Instead, the purpose of this feature is to provide a mechanism to share feedback from the store to help users to decide if an add-on can be useful for them.
```
"keep account" does not have meaning in English. I have adjusted to reach what I believe was your intended meaning.
Also, I have added a line break between the sentences, which translators like.

```suggestion
Please be aware that this doesn't replace direct communication with add-on developers.
Instead, the purpose of this feature is to provide a mechanism to share feedback from the store to help users decide if an add-on may be useful for them.
```

XLTechie commented on user_docs/en/userGuide.t2t:
Thanks for adding this to the user guide @nvdaes.
I have left some comments and suggestions for your consideration, to perhaps help it flow better in English.

XLTechie commented:
@nvdaes, I don't think examples of reviews would be needed. If there need to be 
suggestions for what people can talk about in reviews, this should be on the 
discussion page IMO.

nvdaes committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Luke Davis <8139760+XLTechie@users.noreply.github.com>

XLTechie commented:
@nvdaes, Do you have some unrelated changes in your diff? Like changes to nvWave.py, etc.? GitHub is saying so, the way that I read it.

nvdaes committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Luke Davis <8139760+XLTechie@users.noreply.github.com>

nvdaes committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Luke Davis <8139760+XLTechie@users.noreply.github.com>

nvdaes committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Luke Davis <8139760+XLTechie@users.noreply.github.com>

nvdaes committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Luke Davis <8139760+XLTechie@users.noreply.github.com>

nvdaes commented:
@XLTechie , thanks for your useful suggestion.
About diffs, I tried to merge last changes from master. Is not OK?

XLTechie commented:
@nvdaes Either one of your recent commits fixed it, or there was something odd 
about the GitHub interface. When I checked out the PR locally after your recent commits, 
those extra unrelated files weren't showing as changed.

So all is fine with that now.

nvdaes commented:
@XLTechie wrote: 

> So all is fine with that now.

Great. Thanks for your valuable comments.

XLTechie commented on source/gui/_addonStoreGui/controls/details.py:
```diff
@@ -314,6 +314,13 @@ def _refresh(self):
 						details.sourceURL
 					)
 
+					if details.reviewUrl is not None:
```
Shouldn't the "reviewUrl" variable in all of these places, be capitalized the same as "details.sourceURL"?

nvdaes commented:
> Shouldn't the "reviewUrl" variable in all of these places, be capitalized the same as "details.sourceURL"?

I think no since in metadata the variable is set as reviewUrl, and I didn't realise about capitalization, and discussions are created for some add-ons at this moment, and in jsonschema of addon-datastore-validation the variable is also named reviewUrl.

XLTechie commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -241,6 +241,21 @@ def _makeActionsList(self):
 			),
+			AddonActionVM(
+				# Translators: Label for an action that opens the webpage to see and send feedback for the selected add-on
+				displayName=pgettext("addonStore", "See or provide &feedback"),
```
My same comment here as in the documentation: in my opinion this should be "Read", not "See". Or, the entire option should just be called "Reviews"--this menu tends to use short options.

seanbudd replied:
I agree - this should just be "Reviews"

CyrilleB79 replied:
@seanbudd and @XLTechie, "Review" is not suitable IMO:
* "Review" may let think to a developer's review. This is more true for add-ons given reviewing add-ons was part of the process to publish add-ons on the old website. This can lead to confusion.
* At least in French, I would have difficulty to translate "Review" so that people understand that it is for anybody.

I would prefer the longer but more explicit "Read and provide feedback". But if you stick to something shorter, "Feedback" seems more suitable than "Review".

Of course, an alternative is also to merge the PR as is and wait for feedback from alpha users and translators. The UG and the GUI can be changed later if a more appropriate term is found.

nvdaes replied:
I agree with Cyrille: reviews sounds like the technical procedure for the old website, and feedback is mor understandable, imo, at least in certain languages; anyway imo this shouldn"t prervent to merge this pr.

XLTechie commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -241,6 +241,21 @@ def _makeActionsList(self):
+				actionHandler=lambda aVM: startfile(
+					cast(
+						str,
+						cast(_AddonStoreModel, aVM.model).reviewUrl
```
See my previous suggestion about changing the capitalization of these to "reviewURL", for consistency with other "*URL" variables and fields.

XLTechie commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -241,6 +241,21 @@ def _makeActionsList(self):
+				),
+				validCheck=lambda aVM: (
+					isinstance(aVM.model, _AddonStoreModel)
+					and aVM.model.reviewUrl is not None
```
And here.

nvdaes committed:
Address review: use read instead of see for feedback

nvdaes committed:
Adress review: use reviewURL instead of reviewURL

nvdaes commented:
@XLTechie I think I've addressed all your comments. Imo, we should use the word feedback, since review may sound as something technical, and feedback seems more user friendly.

nvdaes commented:
Also, note that see feedback is used in comments for translators since see is shorter than read.

seanbudd commented on source/braille.py:
```diff
@@ -373,15 +374,20 @@
-		return obj._hasNavigableText
+	if obj._hasNavigableText or isinstance(obj, EditableText):
+		return True
+	return False
```
these changes should probably be removed in the diff - I imagine merging master in will fix this

seanbudd commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -241,6 +241,21 @@ def _makeActionsList(self):
 			),
+			AddonActionVM(
+				# Translators: Label for an action that opens the webpage to see and send feedback for the selected add-on
+				displayName=pgettext("addonStore", "Read or provide &feedback"),
```
Please add an appropriate accelerator key, but this should just be "Reviews"
```suggestion
				displayName=pgettext("addonStore", "Reviews"),
```

seanbudd commented on source/nvwave.py:
```diff
@@ -840,7 +845,11 @@ def __del__(self):
+			# reference to this instance by itself. We don't need to do it explicitly
+			# here. Furthermore, doing it explicitly might cause an exception because
+			# a weakref callback can run before __del__ in some cases, which would mean
+			# it has already been removed from _instances.
```
same here re merging master to remove this diff

seanbudd commented on source/touchHandler.py:
```diff
@@ -281,8 +281,8 @@ def pump(self):
-			# Ensure we are pumpped again by the time more pending multiTouch trackers are ready
-			self.pendingEmitsTimer.Start(interval*1000,True)
+			# Ensure we are pumped again by the time more pending multiTouch trackers are ready
+			self.pendingEmitsTimer.Start(int(interval * 1000), True)
```
same here re merging master to remove this diff

seanbudd reviewed (commented):
There are a large amount of irrelevant changes in this diff. I will re-review from the start once these are resolved.

nvdaes committed:
Address review: use Reviews for the adction menu

nvdaes commented:
@seanbudd , I've addressed your reviews and removed irrelevant diffs.

seanbudd commented on source/gui/_addonStoreGui/controls/details.py:
```diff
@@ -314,6 +314,13 @@ def _refresh(self):
+					if details.reviewURL is not None:
+						self._appendDetailsLabelValue(
+							# Translators: Label for an extra detail field for the selected add-on. In the add-on store dialog.
+							pgettext("addonStore", "Review URL:"),
```
```suggestion
							pgettext("addonStore", "Reviews URL:"),
```

seanbudd commented on source/gui/_addonStoreGui/viewModels/store.py:
```diff
@@ -241,6 +241,21 @@ def _makeActionsList(self):
 			),
+			AddonActionVM(
+				# Translators: Label for an action that opens the webpage to see and send feedback for the selected add-on
+				displayName=pgettext("addonStore", "Re&views"),
```
Maybe this could be "Community reviews" to clear the misconception surrounding developer reviews

XLTechie replied:
"User reviews" might be a bit more standard.

I do prefer some usage of "reviews" over "feedback", however--Feedback suggests an interaction between the users/community and the author. Reviews suggest comments meant to be read by other users, at least in the way it's utilized by Amazon, Ebay, Google, etc.

nvdaes replied:
I prefer Community reviews, so that everybody (users, but also developers and others) are encouraged to participate.

seanbudd reviewed (commented):
I'll review the user guide once we settle on a label for the review action

nvdaes committed:
Update source/gui/_addonStoreGui/controls/details.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

nvdaes committed:
Use Community reviews for the label of reviews action

seanbudd commented on user_docs/en/userGuide.t2t:
```diff
@@ -2840,6 +2840,14 @@ If the add-on was previously "disabled", the status will show "enabled after res
++++ Reviewing add-ons and reading reviews +++[AddonStoreReviews]
+Before installing an add-on, you may want to read opinions and reactions by others.
+Also, it is helpful to other users, if you  take time to provide feedback about add-ons you have tried.
+To see or share your feedback about add-ons, select an add-on and use the See or provide feedback option in the actions menu.
```
```suggestion
To read reviews for an add-on, select an add-on and use the "Community reviews" action.
```

nvdaes committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

nvdaes commented:
@seanbudd I think I've applied all your suggestions for the user guide. I've used the batch option.

seanbudd reviewed (approved):
Thanks @nvdaes

nvdaes commented:
Thank you @sean for approving this, and thanks to other reviewers.
I'll request feedback in the add-ons mailing list and in the spanish community when this is merged.

Qchristensen reviewed (approved):
This will be a well received feature I think.  Good work everyone
